fix: existing ubuntu user fix, and allow Docker run to execute as root for backward compatibility

keivenchang opened the pull request:
#### Overview:

Fix dev target user creation to work correctly in variable environments (e.g. non-local) where existing users/groups may conflict with specified UID/GID values.

#### Details:

Below is a matrix of `docker build` and `docker run` building/executing as different users.
```
+----------------------+----------------------+----------------------+----------------------+
|                      | A: host-user BUILD   | B: root-user BUILD   | C: non-existing-user |
|                      |                      |                      | BUILD                |
+----------------------+----------------------+----------------------+----------------------+
| 1: host-user RUN     | A1 (A+1): PERFECT    | B1 (B+1): ISSUES     | C1 (C+1): PARTIAL    |
|                      | Full R/W access      | Can't write container| Write mounts only    |
|                      | user-ids match       | dirs owned by root   | Can't write container|
+----------------------+----------------------+----------------------+----------------------+
| 2: root-user RUN     | A2 (A+2): SECURITY   | B2 (B+2): WORKS      | C2 (C+2): SECURITY   |
|                      | RISK                 | Expected behavior    | RISK                 |
|                      | Unnecessary root     | (still risky)        | Unnecessary root     |
+----------------------+----------------------+----------------------+----------------------+
| 3: non-existing-user | A3 (A+3): NO ACCESS  | B3 (B+3): ISSUES     | C3 (C+3): NO MOUNT   |
| RUN                  | Can't write anywhere | Can't write container| ACCESS               |
|                      | user-id mismatch     | dirs owned by root   | Mounts owned by host |
+----------------------+----------------------+----------------------+----------------------+
```
- The combination of C+2 (non-existing user 1000 BUILD + root-user RUN) was what we had before.
- I'm turning this into A+1 (host-user BUILD + host-user RUN) for vLLM
- However, I also understand that we should have backward compatibility, so I've added the capability such that you can do `build.sh --uid 1000 --gid 1000` and `run.sh --user root ...`, to achieve exactly what we had before (C+2).
- Handle existing GID conflicts by reusing the group if it already exists instead of failing
- Fix empty USER_UID/USER_GID build arguments that caused useradd failures in environments with existing ubuntu users
- Add debug output to show UID/GID values being used during build for easier troubleshooting
- Ensure consistent Docker image tagging whether --target dev is specified explicitly or used as default
- Add Dockerfile.test for isolated testing of user creation logic

#### Where should the reviewer start?

- container/Dockerfile.vllm - See the improved user creation logic (lines 344-351)
- container/build.sh - Check the UID/GID handling and debug output (lines 496-505)


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

- New Features
  - Standardized non-root user setup in development images with configurable UID/GID, home, and workspace.
  - Enhanced developer environment: user-scoped PATH/LD_LIBRARY_PATH, cargo/rust toolchain dirs, and PYTHONPATH.
  - Added maturin installation for per-user builds and quality-of-life shell/history defaults.
- Tests
  - Introduced a minimal test image that provisions a configurable non-root user and verifies identity at runtime.
- Chores
  - Added build-time logging to display resolved USER_UID and USER_GID when building development images.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

keivenchang committed:
fix: improve dev target user creation for computelab compatibility

- Handle existing GID conflicts by reusing groups
- Fix empty USER_UID/USER_GID build arg issues in computelab environments
- Add debug output for UID/GID values during build
- Ensure consistent tagging whether --target dev is explicit or default
- Add Dockerfile.test for debugging user creation issues
- Make build work on both local and computelab environments

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

oandreeva-nv reviewed (approved):
worked for me for container building

nnshah1 commented on container/Dockerfile.vllm:
```diff
@@ -333,20 +336,20 @@ RUN apt-get update -y && \
+# Create non-root user with host UID/GID for proper file permissions. This ensures
+# files created in mounted volumes have correct ownership. If an ubuntu user already
+# exists, it's removed first and recreated with the correct UID/GID. When the specified
+# GID conflicts with an existing group, that group is reused instead of creating a new one.
```
I think its not a bad idea to create a non root user and make that the default - what I don't quite understand is if that should match the user during build time. 

IIUC VS code and others will update the UID / GUID when it is used - we can do the same during --mount-workspace if we want - so if you mount we also pass --USER uid and --GID gid during docker run - 

I think the default flow should build ideally the same image independent of where it is built -

keivenchang replied:
You can execute `docker run --user <any uid>:<any gid>`, but now the problem is that inside the container, it won't be able to access 1) the ~ubuntu/.bashrc 2) the RUSTUP_HOME and CARGO_HOME, which contain not just the Rust/Cargo binaries, but the caches that Cargo/Rust use. We we used to run as root, this was never a problem -- root could clobber any file.

FYI, for experimentation, I just built an image with a non-existent uid:gid (12345/54321), then `docker run --user <real-uid>:<real-gid> ...` and confirmed that inside the container, the real-user ID cannot write to files owned by a completely different uid:gid.

Below is a matrix of `docker build` and `docker run` possibilities.
```
+----------------------+----------------------+----------------------+----------------------+
|                      | A: host-user Build   | B: root-user Build   | C: non-exist-user-id |
|                      |                      |                      | Build                |
+----------------------+----------------------+----------------------+----------------------+
| 1: host-user Run     | A1: PERFECT          | B1: ISSUES           | C1: PARTIAL          |
|                      | user-ids match       | host-user can't write| Different user-ids   |
|                      |                      | root-user files      |                      |
+----------------------+----------------------+----------------------+----------------------+
| 2: root-user Run     | A2: WORKS            | B2: WORKS            | C2: WORKS            |
|                      | root-user overrides  | root-user to root-   | root-user bypasses   |
|                      |                      | user                 |                      |
+----------------------+----------------------+----------------------+----------------------+
| 3: non-exist-user-id | A3: LIMITED          | B3: ISSUES           | C3: MATCH ONLY       |
| Run                  | user-id mismatch     | non-exist-user       | Works if same        |
|                      |                      | limited              | user-id              |
+----------------------+----------------------+----------------------+----------------------+
```

The combination of C+2 (non-existing user 1000 + root-user execution) was what we had before. I'd like to turn this into A+1 (host-user build + host-user execution). However, I also understand that we should have backward compatibility, so I've added the capability such that you can do `build.sh --uid 1000 --gid 1000` and `run.sh --user root ...`, to achieve exactly what we had before (C+2).

I hope this is diagram makes it clearer -- it took me some time to think about how to present all the different scenarios.

keivenchang committed:
Fix Dockerfile.vllm user permissions for computelab environment

- Update Dockerfile.vllm to handle existing ubuntu user scenarios
- Modify build.sh to support UID/GID configuration for proper volume mount permissions

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang committed:
feat: add --user parameter and improve run.sh user handling

- Add --user parameter for explicit container user control
- Make --privileged independent of --mount-workspace
- Only pass HF_TOKEN/GITHUB_TOKEN if set in environment
- Fix user permission issues with mismatched UIDs

Signed-off-by: Keiven Chang <keivenchang@users.noreply.github.com>

keivenchang commented on container/build.sh:
```diff
@@ -500,6 +500,7 @@ if [[ $TARGET == "dev" ]]; then
     if [ -z "$USER_GID" ]; then
         USER_GID=$(id -g)
     fi
+    echo "Building dev target with USER_UID=$USER_UID USER_GID=$USER_GID"
```
FYI, 1) You can still `build.sh --uid 1000 --gid 1000` to preserve the old behavior. Otherwise, it'll default to the host's user 2) I added a `run.sh --user root ...` param, so that you can execute as root, just as before, if you want to preserve the exact same behavior.
